### ISSUE TITLE
Vim9: unexpected error when using interpolated string in `:def` function

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#10377]
**Steps to reproduce**

Run this shell command:

    vim -Nu NONE -S <(tee <<'EOF'
        vim9script
        def Func(): list<bool>
            var s = 'x'
            return 'axb foo axb bar axb baz'
                ->split($'a{s}b')
                ->map((_, v: string) => v =~ 'bar')
        enddef
        echo Func()
    EOF
    )

An error is given:

    E1013: Argument 2: type mismatch, expected func(?number, ?string): string but got func(any, string): bool

**Expected behavior**

No error is given.  This is echo'ed:

    [false, true, false]

**Version of Vim**

8.2 Included patches: 1-4912

**Environment**

Operating system: Ubuntu 20.04.4 LTS
Terminal: xterm
Value of $TERM: xterm-256color
Shell: GNU bash, version 5.0.17

**Additional context**

At the script level, the exact same code does not give any error:
```vim
vim9script
var s = 'x'
echo 'axb foo axb bar axb baz'
    ->split($'a{s}b')
    ->map((_, v: string) => v =~ 'bar')
```
    [false, true, false]

---

In compiled code, if we replace the interpolated string:

    ->split($'a{s}b')

with a concatenation:

    ->split('a' .. s .. 'b')

Then, no error is given:
```vim
vim9script
def Func(): list<bool>
    var s = 'x'
    return 'axb foo axb bar axb baz'
        ->split('a' .. s .. 'b')
        ->map((_, v: string) => v =~ 'bar')
enddef
echo Func()
```
    [false, true, false]

**Comment by brammool** [comment]
> **Steps to reproduce**
> 
> Run this shell command:
> 
>     vim -Nu NONE -S <(tee <<'EOF'
>         vim9script
>         def Func(): list<bool>
>             var s = 'x'
>             return 'axb foo axb bar axb baz'
>                 ->split($'a{s}b')
>                 ->map((_, v: string) => v =~ 'bar')
>         enddef
>         echo Func()
>     EOF
>     )
> 
> An error is given:
> 
>     E1013: Argument 2: type mismatch, expected func(?number, ?string): string but got func(any, string): bool
> 
> **Expected behavior**
> 
> No error is given.  This is echo'ed:
> 
>     [false, true, false]

It works OK when joining the lines:

             return 'axb foo axb bar axb baz'
                 ->split($'a{s}b')->map((_, v: string) => v =~ 'bar')

Problem is that evaluating "s" looks in the next line.


-- 
hundred-and-one symptoms of being an internet addict:
136. You decide to stay in a low-paying job teaching just for the
     free Internet access.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by LemonBoy** [comment]
Fixed by #10378
